File: Python/53.Maximum Subarray.py
#寻找和最大的子串
class Solution(object):
    def maxSubArray(self, nums):
        ans, sum = nums[0], nums[0]
        for i in range(1, len(nums)):
            if (sum < 0):
                sum = nums[i]
            else:
                sum += nums[i]
            ans = max(ans, sum)
        return ans

s = Solution()
print(s.maxSubArray([-2,1,-3,4,-1,2,1,-5,4]))
print(s.maxSubArray([1]))
print(s.maxSubArray([-5,-3,-1,-2]))

File: C++/Array/53.Maximum Subarray.cpp
/*
求数组的一个子数组，和最大。
*/
/*
思路：用前面的最大和加当前数，与当前数比较，大的作为最大和。
 */

#include <iostream>
#include <vector>
#include <windows.h>
#include <windef.h>
#include <limits.h>
using namespace std;

class Solution {
public:
	int maxSubArray(vector<int>& nums) {
        int largestsum = INT_MIN,current;
        if(nums.size()==0)	return largestsum;
        largestsum = current = nums[0];
        for(int i=1; i < nums.size(); ++i)
        {
        	current = max(nums[i],current+nums[i]);
        	largestsum = max(current,largestsum);
        }
    	return largestsum;
    }
};

int main()
{
	Solution s;
	vector<int> nums{-2,1,-3,4,-1,2,1,-5,4};
	cout<<s.maxSubArray(nums)<<endl;
}

